trade-finance-manager-ui/server/controllers/login/loginService.js

```diff
@@ -0,0 +1,72 @@
+const { SSO } = require("../../constants");
+
+/**
+ * verifyReferrerForExternalSsoPost
+ * Verify referrer, allow empty referrer for localhost.
+ * 1) If host is localhost then referrer will be missing, no need to check it
+ * 2) Return error message if referrer is missing or it is not from SSO authority
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
```

Comment:
I think it'd be a nice pattern to throw errors as errors as opposed to returning strings with the error message. I think this would keep our types cleaner, as well as making errors easier to handle. What do you think? Could we look at doing it for `verifyReferrerForExternalSsoPost` at least? I find the idea this returns `false` when it is verified is surprising!

Reply:
Agree, didn't spot this before 

@avaitonis How can we determine a correct `referrer` here? Maybe could use an existing env var for TFM UI? i'm thinking we could change these functions to be simpler and `verifyReferrerForExternalSsoPost` has a more logical return type like @AlexBramhill mentioned, e.g (simple example)

```js
const verifyReferrerForExternalSsoPost = (req) => {
  if (referrer !=== VALID_REFERRER) {
    // one of these...
    // return false;
    // throw new Error();
  }

  return true;
};

// ...

const acceptExternalSsoPost = () => {
  const validReferrer = verifyReferrerForExternalSsoPost();

  if (validReferrer) {
    return res.render('sso/accept-external-sso-post.njk', templateParams);
  }

  return res.status(500).render('_partials/problem-with-service.njk');
};
```

Reply:
one reason i like throwing errors here is that it allows us to see the happy path in one block, and also keeps our responses all at the bottom -- ie
```
const verifyReferrerForExternalSsoPost = (referrer) => {  
  if (referrer !=== VALID_REFERRER) {  
    throw new Error();  
  }  
  // notice no return here
};  

// ...  

const acceptExternalSsoPost = (req) => {  
  try {
    verifyReferrerForExternalSsoPost(req.referrer);  
    verifyBodyForExternalSsoPost(req.body);  

    return res.render('sso/accept-external-sso-post.njk', templateParams);  
  } catch (error) {
    return res.status(500).render('_partials/problem-with-service.njk', { error: { message: error }});
  }
};  
```

Additional points for custom errors but we can add these later...

Reply:
Thanks @AlexBramhill I like this, only concern with `verifyReferrerForExternalSsoPost` is that IMO it should always return "something", not sure if this is the case in DTFS/TFM API but there is usually a linting rule that a function should always return something

Reply:
Changed to Alex code.

Verification function never return anything, so lint is happy with this.